Libcxx tests failed. I'll check them tomorrow.
```
Failed Tests (15):
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/extents/assert.conversion.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/extents/assert.ctor_from_array.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/extents/assert.ctor_from_integral.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/extents/assert.ctor_from_span.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/extents/assert.obs.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_left/assert.conversion.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_left/assert.ctor.extents.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_left/assert.ctor.layout_right.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_left/assert.ctor.layout_stride.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_right/assert.conversion.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_right/assert.ctor.extents.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_right/assert.ctor.layout_left.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_right/assert.ctor.layout_stride.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_stride/assert.ctor.extents_array.pass.cpp
  llvm-libc++-shared.cfg.in :: libcxx/containers/views/mdspan/layout_stride/assert.ctor.extents_span.pass.cpp
```

```
# | /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-bt7k7-1/llvm-project/github-pull-requests/libcxx/test/libcxx/containers/views/mdspan/extents/assert.conversion.pass.cpp:39:29: error: declaration of 'dynamic_extent' must be imported from module 'std.span.fwd' before it is required
# |    39 |   constexpr size_t D = std::dynamic_extent;
# |       |                             ^
# | /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-bt7k7-1/llvm-project/github-pull-requests/build/runtimes/runtimes-bins/libcxx/test-suite-install/include/c++/v1/__fwd/span.h:28:25: note: declaration here is not visible
# |    28 | inline constexpr size_t dynamic_extent = numeric_limits<size_t>::max();
# |       |                         ^
# | 1 error generated.
```

--- a/clang/lib/AST/Decl.cpp
+++ b/clang/lib/AST/Decl.cpp
@@ -400,9 +400,9 @@ void LinkageComputer::mergeTemplateLV(
   FunctionTemplateDecl *temp = specInfo->getTemplate();
   // Merge information from the template declaration.
   LinkageInfo tempLV = getLVForDecl(temp, computation);
-  // The linkage of the specialization should be consistent with the
-  // template declaration.
-  LV.setLinkage(tempLV.getLinkage());
+  // The linkage and visibility of the specialization should be
+  // consistent with the template declaration.
+  LV.mergeMaybeWithVisibility(tempLV, considerVisibility);
 
   // Merge information from the template parameters.
   LinkageInfo paramsLV =
@@ -1051,6 +1051,13 @@ LinkageComputer::getLVForClassMember(const NamedDecl *D,
     if (const auto *redeclTemp = dyn_cast<RedeclarableTemplateDecl>(temp)) {
       if (isExplicitMemberSpecialization(redeclTemp)) {
         explicitSpecSuppressor = temp->getTemplatedDecl();
+      } else if (const RedeclarableTemplateDecl *from =
+                     redeclTemp->getInstantiatedFromMemberTemplate()) {
+        // If no explicit visibility is specified yet, and this is an
+        // instantiated member of a template, look up visibility there
+        // as well.
+        LinkageInfo fromLV = from->getLinkageAndVisibility();
+        LV.mergeMaybeWithVisibility(fromLV, considerVisibility);
       }
     }
   }

--- a/clang/test/CodeGenCXX/visibility.cpp
+++ b/clang/test/CodeGenCXX/visibility.cpp
@@ -1457,9 +1457,45 @@ namespace test71 {
   // CHECK-LABEL: declare hidden noundef i32 @_ZN6test713fooIiE3zedEv(
   // CHECK-LABEL: define linkonce_odr noundef i32 @_ZN6test713fooIiE3barIiEET_v(
   // CHECK-LABEL: define linkonce_odr hidden noundef i64 @_ZN6test713fooIlE3zedEv(
-  // CHECK-LABEL: define linkonce_odr noundef i32 @_ZN6test713fooIlE3barIiEET_v(
+  // CHECK-LABEL: define linkonce_odr hidden noundef i32 @_ZN6test713fooIlE3barIiEET_v(
   // CHECK-HIDDEN-LABEL: declare hidden noundef i32 @_ZN6test713fooIiE3zedEv(
   // CHECK-HIDDEN-LABEL: define linkonce_odr noundef i32 @_ZN6test713fooIiE3barIiEET_v(
   // CHECK-HIDDEN-LABEL: define linkonce_odr hidden noundef i64 @_ZN6test713fooIlE3zedEv(
   // CHECK-HIDDEN-LABEL: define linkonce_odr hidden noundef i32 @_ZN6test713fooIlE3barIiEET_v(
+}
+
+// https://github.com/llvm/llvm-project/issues/103477
+namespace test72 {
+  template <class a>
+  struct t {
+    template <int>
+    static HIDDEN void bar() {}
+  };
+
+  void test() {
+      t<char>::bar<1>();
+  }
+  // CHECK-LABEL: define linkonce_odr hidden void @_ZN6test721tIcE3barILi1EEEvv(
+  // CHECK-HIDDEN-LABEL: define linkonce_odr hidden void @_ZN6test721tIcE3barILi1EEEvv(
+}
+
+// https://github.com/llvm/llvm-project/issues/31462
+namespace test73 {
+  template <class T> struct s {
+    template <class U>
+    __attribute__((__visibility__("hidden"))) U should_not_be_exported();
+  };
+
+  template <class T> template <class U> U s<T>::should_not_be_exported() {
+    return U();
+  }
+
+  extern template struct __attribute__((__visibility__("default"))) s<int>;
+
+  int f() {
+    s<int> o;
+    return o.should_not_be_exported<int>();
+  }
+  // CHECK-LABEL: define linkonce_odr noundef i32 @_ZN6test731sIiE22should_not_be_exportedIiEET_v(
+  // CHECK-HIDDEN-LABEL: define linkonce_odr noundef i32 @_ZN6test731sIiE22should_not_be_exportedIiEET_v(
 }

